### ISSUE TITLE
Simple example failure on big-endian systems

### ISSUE THREAD

**Issue by ginggs** [kaskr/adcomp#340]
Continued from https://github.com/glmmTMB/glmmTMB/issues/665

#### Description:
Since Matrix was upgraded to 1.3-0, and continuing with 1.3-2, the simple example from TMB fails with:
```
Error in nlminb(obj$par, obj$fn, obj$gr) :
  gradient function must return a numeric vector of length 4
```

#### Reproducible Steps:
install `r-base-core` and `r-base-dev` on a big-endian system (e.g. s390x, ppc64) running Debian testing, run R.
```
> install.packages("TMB")
...
> library(TMB)
> runExample("simple")
Running example simple
...
> ## Fit model
> obj <- MakeADFun(data=list(x=x, B=B, A=A),
+                  parameters=list(u=u*0, beta=beta*0, logsdu=1, logsd0=1),
+              .... [TRUNCATED]

> opt <- nlminb(obj$par, obj$fn, obj$gr)
Error in nlminb(obj$par, obj$fn, obj$gr) :
  gradient function must return a numeric vector of length 4
In addition: Warning message:
In nlminb(obj$par, obj$fn, obj$gr) : NA/NaN function evaluation
```
```
> traceback()
6: nlminb(obj$par, obj$fn, obj$gr) at simple.R#27
5: eval(ei, envir)
4: eval(ei, envir)
3: withVisible(eval(ei, envir))
2: source(paste(name, ".R", sep = ""), echo = TRUE)
1: runExample("simple")
```

#### TMB Version:
```
> packageVersion("TMB")
[1] '1.7.19'
```

#### R Version:
```
> R.version.string
[1] "R version 4.0.4 (2021-02-15)"
```

#### Operating System:
Debian bullseye (testing)

**Comment by kaskr** [comment]
I've confirmed this issue by running the image
https://cloud-images.ubuntu.com/bionic/current/bionic-server-cloudimg-s390x.img
using `qemu`.

With old `Matrix` package (1.2-12) there are no problems.
With new `Matrix` package (1.3-2) the function `TMB:::updateCholesky()` always returns `FALSE`, indicating (wrongly) that matrix is not PD.

Digging a bit deeper I found that after calling (as done [here](https://github.com/kaskr/adcomp/blob/597fad47af9c0b22ab69d2201272a8e865c7e451/TMB/src/solve_subset.c#L130) )

```
M_cholmod_factorize_p(A, mm, (int*)NULL, 0 /*fsize*/, f, &c)
```

the `cholmod_common` struct contains bogus, e.g. `c.status=14903956` and `c.dtype=15216988`.

I still don't know if it's really a big-endian issue or something else...

**Comment by kaskr** [comment]
@ginggs Thanks for the report! Could you please check if the `big_endian` branch solves it? (it shouldn't be necessary to re-install other packages than TMB to check it...)

**Comment by ginggs** [comment]
@kaskr the `big_endian` branch does indeed fix the test failure I was seeing in glmmTMB.

I was able to reproduce the original issue on two different 64-bit big-endian architectures, but not on a 32-bit big-endian architecture, this makes me think somewhere there's a 64-bit variable that's being accessed as a 32-bit variable in a different part of the code.

So I don't think it's an endianness issue as such, in the way that data is being written in one byte order and read in another.  Rather, it's a bug that just shows up more easily on a 64-bit big-endian architecture.

**Comment by bbolker** [comment]
Just curious, can you explain what/how this fixed it? 232ef2ee

```
 int ok = (f->minor == f->n); // WAS: (c.status == CHOLMOD_OK);
```

**Comment by kaskr** [comment]
@bbolker According to the [CHOLMOD userguide](https://github.com/DrTimothyAldenDavis/SuiteSparse/blob/master/CHOLMOD/Doc/CHOLMOD_UserGuide.pdf) there are at least two ways to check the PD status of a numerical factorization. One uses the `cholmod_common` struct (`c->status=CHOLMOD_OK`) the other uses the `cholmod_factor` struct (`f->minor=f->n`).
The former no longer behaves as documented while the latter still seems to work. It would be nice to understand what's going on but I'm happy with the quick fix for now.
